### PR TITLE
udp_session_filters: defer socket creation for after iterating filters

### PR DESCRIPTION
**Pull request by ohadvano**
Additional Description:
This is a minor behavior change that only affects paths that uses UDP session filters (added in #29366). Without this PR, choosing the upstream host and socket creation happens right when the ``ActiveSession`` starts. This limits the session filters abilities to dynamically affect the chosen host (for example, for DFP scenarios) and may cause creating sockets and exhausting ports even though session filters may ``StopIteration`` if the session needs to be dynamically terminated.
The new behavior will only choose the host and create the socket after all session filters iterated ``onNewSession()``. If there are no filters applied, this PR has no side-effect.
Additionally, since per-packet load balancing requires choosing the host in advance, an exception will be thrown if ``per_packet_load_balancing`` is used in conjunction with ``session_filters``.
Additionally2, since per-datagram host health-checks also require choosing the host in advance, if the UDP proxy has session filters, these health checks will not be performed.
I believe that generally this would be guarded by runtime flag, but since the session filters were only merged few days ago and the previous behavior kind of doesn't fully makes sense, I believe we could and should not use runtime flag here.

Risk Level: low
Testing: Integration tests
Docs Changes: Changelog, UDP proxy docs
Release Notes: None
Platform Specific Features: None

### CODE REVIEW

**Comment by ohadvano** [comment]
/unassign @lizan 
/assign @mattklein123 
Thanks again for the reviews, can you please take a look?

**Comment by ohadvano** [comment]
@mattklein123 can you please take a look?

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc [hunk 1]
```diff
@@ -178,25 +191,30 @@ UdpProxyFilter::StickySessionClusterInfo::StickySessionClusterInfo(
...
       return Network::FilterStatus::StopIteration;
     }
   } else {
     active_session = active_session_it->get();
-    if (active_session->host().coarseHealth() == Upstream::Host::Health::Unhealthy) {
+    // We defer the socket creation when the session includes filters, so the filters can be
+    // iterated before choosing the host, to allow dynamically choosing upstream host. Due to this,
+    // we can't perform health checks during a session.
+    if (!defer_socket &&
+        active_session->host().value().get().coarseHealth() == Upstream::Host::Health::Unhealthy) {
```
Comment (mattklein123) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:208]: Wouldn't you want this behavior to happen at the end of the session filter chain though? Or give filters the chance to interact with this behavior? This seems like a regression in functionality for those cases. At minimum I would add some kind of TODO about this?

/wait

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with 1 question, thanks.

/wait

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc (hunk 1 above)
Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc]: I see your point. It makes sense that a similar check can also happen when using filters as you suggested. I added a TODO and will add this in a separate PR, need to do some more thinking about this. Thanks

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**